File: src/main/java/org/moinex/model/investment/BondInterestCalculation.java
/*
 * Filename: BondInterestCalculation.java
 * Created on: February 20, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.model.investment;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.YearMonth;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import org.moinex.util.Constants;

@Entity
@Table(name = "bond_interest_calculation")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder
public class BondInterestCalculation {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    @Setter(AccessLevel.NONE)
    private Integer id;

    @ManyToOne(optional = false)
    @JoinColumn(name = "bond_id", referencedColumnName = "id", nullable = false)
    private Bond bond;

    @Column(name = "reference_month", nullable = false)
    private String referenceMonth;

    @Column(name = "calculation_date", nullable = false)
    private String calculationDate;

    @Column(name = "quantity", nullable = false)
    private BigDecimal quantity;

    @Column(name = "invested_amount", nullable = false)
    private BigDecimal investedAmount;

    @Column(name = "monthly_interest", nullable = false)
    private BigDecimal monthlyInterest;

    @Column(name = "accumulated_interest", nullable = false)
    private BigDecimal accumulatedInterest;

    @Column(name = "final_value", nullable = false)
    private BigDecimal finalValue;

    @Column(name = "calculation_method")
    private String calculationMethod;

    @Column(name = "created_at", nullable = false)
    private String createdAt;

    public abstract static class BondInterestCalculationBuilder<
            C extends BondInterestCalculation, B extends BondInterestCalculationBuilder<C, B>> {
        public B referenceMonth(YearMonth referenceMonth) {
            this.referenceMonth = referenceMonth.toString();
            return self();
        }

        public B calculationDate(LocalDate calculationDate) {
            this.calculationDate = calculationDate.format(Constants.DATE_FORMATTER_NO_TIME);
            return self();
        }

        public B createdAt(LocalDateTime createdAt) {
            this.createdAt = createdAt.format(Constants.DB_DATE_FORMATTER);
            return self();
        }
    }

    public YearMonth getReferenceMonth() {
        if (referenceMonth == null) {
            return null;
        }
        return YearMonth.parse(referenceMonth);
    }

    public void setReferenceMonth(YearMonth month) {
        if (month == null) {
            this.referenceMonth = null;
            return;
        }
        this.referenceMonth = month.toString();
    }

    public LocalDate getCalculationDate() {
        if (calculationDate == null) {
            return null;
        }
        return LocalDate.parse(calculationDate, Constants.DATE_FORMATTER_NO_TIME);
    }

    public void setCalculationDate(LocalDate date) {
        if (date == null) {
            this.calculationDate = null;
            return;
        }
        this.calculationDate = date.format(Constants.DATE_FORMATTER_NO_TIME);
    }

    public LocalDateTime getCreatedAt() {
        if (createdAt == null) {
            return null;
        }
        return LocalDateTime.parse(createdAt, Constants.DB_DATE_FORMATTER);
    }

    public void setCreatedAt(LocalDateTime dateTime) {
        if (dateTime == null) {
            this.createdAt = null;
            return;
        }
        this.createdAt = dateTime.format(Constants.DB_DATE_FORMATTER);
    }
}

File: src/main/kotlin/org/moinex/model/investment/BondInterestCalculation.kt
/*
 * Filename: BondInterestCalculation.kt (original filename: BondInterestCalculation.java)
 * Created on: February 20, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 07/03/2026
 */

package org.moinex.model.investment

import jakarta.persistence.Column
import jakarta.persistence.Convert
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import org.moinex.common.converter.LocalDateStringConverter
import org.moinex.common.converter.LocalDateTimeStringConverter
import org.moinex.common.converter.YearMonthStringConverter
import org.moinex.common.extension.toRounded
import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.YearMonth

@Entity
@Table(name = "bond_interest_calculation")
class BondInterestCalculation(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Int? = null,
    @ManyToOne(optional = false)
    @JoinColumn(name = "bond_id", referencedColumnName = "id", nullable = false)
    var bond: Bond,
    @Column(name = "reference_month", nullable = false)
    @Convert(converter = YearMonthStringConverter::class)
    var referenceMonth: YearMonth,
    @Column(name = "calculation_date", nullable = false)
    @Convert(converter = LocalDateStringConverter::class)
    var calculationDate: LocalDate,
    @Column(name = "calculated_until_date")
    @Convert(converter = LocalDateStringConverter::class)
    var calculatedUntilDate: LocalDate? = null,
    @Column(name = "quantity", nullable = false)
    var quantity: BigDecimal,
    @Column(name = "invested_amount", nullable = false)
    var investedAmount: BigDecimal,
    @Column(name = "monthly_interest", nullable = false)
    var monthlyInterest: BigDecimal,
    @Column(name = "accumulated_interest", nullable = false)
    var accumulatedInterest: BigDecimal,
    @Column(name = "final_value", nullable = false)
    var finalValue: BigDecimal,
    @Column(name = "calculation_method")
    var calculationMethod: String? = null,
    @Column(name = "created_at", nullable = false)
    @Convert(converter = LocalDateTimeStringConverter::class)
    var createdAt: LocalDateTime,
    @Column(name = "manually_adjusted", nullable = false)
    var manuallyAdjusted: Boolean = false,
) {
    init {
        quantity = quantity.toRounded()
        investedAmount = investedAmount.toRounded()
        monthlyInterest = monthlyInterest.toRounded()
        accumulatedInterest = accumulatedInterest.toRounded()
        finalValue = finalValue.toRounded()

        require(quantity >= BigDecimal.ZERO) {
            "Quantity must be non-negative"
        }
        require(investedAmount >= BigDecimal.ZERO) {
            "Invested amount must be non-negative"
        }
    }

    fun isManuallyAdjusted(): Boolean = manuallyAdjusted

    override fun toString(): String = "BondInterestCalculation [id=$id, bond=${bond.name}, month=$referenceMonth]"
}
